python/lsst/afw/math/stack.cc

```diff
@@ -32,6 +35,31 @@ using namespace py::literals;
 
 using namespace lsst::afw::math;
 
+namespace {
+
+// a variant of statisticsStack for pybind11 that takes vector of vector
+// (instead of vector of shared pointer to vector)
+// and returns a vector instead of a shared pointer to a vector
+template <typename PixelT>
+std::vector<PixelT> _statisticsStack(
+        std::vector<std::vector<PixelT>>& vectors,  ///< Vectors to process
+        Property flags,                                              ///< statistics requested
+        StatisticsControl const& sctrl = StatisticsControl(),        ///< control structure
+        std::vector<lsst::afw::image::VariancePixel> const& wvector =
+                std::vector<lsst::afw::image::VariancePixel>(0)  ///< vector containing weights
+) {
+        std::vector<std::shared_ptr<std::vector<PixelT>>> vcopy;
+        vcopy.reserve(vectors.size());
+        for (auto const & v : vectors) {
+            auto vshared = std::make_shared<std::vector<PixelT>>(std::move(v));
```

Comment:
I think this is unsafe, as it will empty the elements of `vectors`. I realize that the [original statisticsStack](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/namespacelsst_1_1afw_1_1math.html#aa7f9ddedbd14f76864aee051549b5cea) takes the vectors by modifiable reference, but surely that's not the change they're supposed to make.

Reply:
The `vectors` argument is a copy that pybind11 makes, so I figured it was safe to alter it.

What do you suggest instead?

Reply:
If it really is a temporary copy than that's fine, but that's not at all obvious to me. I think at the very least a comment would be necessary.

Reply:
I think the `move` is safe, as long as this function is only called by pybind11.  I'm not sure you're actually moving, though, given that you're calling `std::move` on a const reference; that will just cast the object to `const T&&`, which will invoke the copy constructor.  I think changing the loop variable to `auto & v` should fix that.

All that said, it'd probably be best (and perhaps not even any more disruptive) to just change the main C++ API to take a vector-of-vectors.  Using `shared_ptr` was almost certainly a premature attempt at optimization, and quite possibly a counterproductive one.

Reply:
@TallJimbo Sounds good to me. Any objection to having that function return a std::vector instead of a shared_ptr<std::vector> as well? That would make it directly compatible with pybind11

Reply:
Oh, definitely do that as well.  I didn't realize that was also an issue.  I'd be surprised if we have anywhere in the stack where using a `shared_ptr<vector>` is actually a good idea (which is not to say that fixing this would always be easy, but I think it would be here, because there's probably very little C++ code that uses statisticsStack).

Reply:
OK. I have made that change (including returning a std::vector. It all seems to work and it's certainly an improvement. A global search strongly suggests it won't break anything but I'll start a Jenkins run to be sure.